### PR TITLE
Make primary sort stable

### PR DESCRIPTION
**Pull request by gnusi**
* Change comparator interface to 3-way comparison
* Make primary sort stable

### CODE REVIEW

File: tests/index/assert_format.cpp [hunk 1]
```diff
@@ -392,10 +392,10 @@ void index_segment::sort(const irs::comparer& comparator) {
     return;
   }
 
-  std::sort(sort_.begin(), sort_.end(),
-            [&comparator](const auto& lhs, const auto& rhs) {
-              return comparator(std::get<0>(lhs), std::get<0>(rhs));
-            });
+  std::stable_sort(sort_.begin(), sort_.end(),
+                   [&comparator](const auto& lhs, const auto& rhs) {
```
Comment (MBkkt) [inline tests/index/assert_format.cpp:396]: NIT
```suggestion
                   [&](const auto& lhs, const auto& rhs) {
```

File: core/index/sorted_column.cpp [hunk 2]
```diff
@@ -30,10 +30,17 @@ namespace irs {
...
 bool sorted_column::flush_sprase_primary(
   doc_map& docmap, const columnstore_writer::values_writer_f& writer,
-  doc_id_t docs_count, const comparer& less) {
-  auto comparer = [&less, this](const std::pair<doc_id_t, size_t>& lhs,
-                                const std::pair<doc_id_t, size_t>& rhs) {
-    return less(get_value(&lhs), get_value(&rhs));
+  doc_id_t docs_count, const comparer& compare) {
+  auto comparer = [&compare, this](
+                    const std::pair<doc_id_t, size_t>& lhs,
+                    const std::pair<doc_id_t, size_t>& rhs) -> bool {
```
Comment (MBkkt) [inline core/index/sorted_column.cpp:36]: NIT
```suggestion
  auto comparer = [&](
                    const auto& lhs,
                    const auto& rhs) -> bool {
```

File: core/index/sorted_column.cpp [hunk 3]
```diff
@@ -44,10 +51,10 @@ bool sorted_column::flush_sprase_primary(
...
   std::vector<size_t> sorted_index(index_.size() - 1);
   std::iota(sorted_index.begin(), sorted_index.end(), 0);
-  std::sort(sorted_index.begin(), sorted_index.end(),
-            [&comparer, this](size_t lhs, size_t rhs) {
-              return comparer(index_[lhs], index_[rhs]);
-            });
+  std::stable_sort(sorted_index.begin(), sorted_index.end(),
+                   [&comparer, this](size_t lhs, size_t rhs) {
+                     return comparer(index_[lhs], index_[rhs]);
+                   });
```
Comment (MBkkt) [inline core/index/sorted_column.cpp:57]:
```suggestion
  std::sort(sorted_index.begin(), sorted_index.end(),
                   [&](auto lhs, auto rhs) {
                     return comparer(index_[lhs], index_[rhs]);
                   });
```

File: tests/index/sorted_column_test.cpp [hunk 4]
```diff
@@ -35,26 +35,35 @@
 namespace {
 
 struct Comparator final : irs::comparer {
-  bool less(irs::bytes_view lhs, irs::bytes_view rhs) const noexcept override {
+  int compare(irs::bytes_view lhs,
+              irs::bytes_view rhs) const noexcept override {
```
Comment (MBkkt) [inline tests/index/sorted_column_test.cpp:39]:
```suggestion
              irs::bytes_view rhs) const noexcept final {
```

**Review by MBkkt — APPROVED** [review]

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/index/sorted_column.cpp [hunk 5]
```diff
@@ -28,15 +28,18 @@
...
+  doc_id_t docs_count, const comparer& compare) {
+  auto comparer = [&](const std::pair<doc_id_t, size_t>& lhs,
+                      const std::pair<doc_id_t, size_t>& rhs) -> int {
+    return compare(get_value(&lhs), get_value(&rhs));
   };
 
-  if (std::is_sorted(index_.begin(), index_.end() - 1, comparer)) {
+  if (std::is_sorted(index_.begin(), index_.end() - 1,
+                     [&](const auto& lhs, const auto& rhs) {
+                       return comparer(lhs, rhs) < 0;
```
Comment (MBkkt) [inline core/index/sorted_column.cpp:41]: add assert

**Review by MBkkt — APPROVED** [review]

**Pull request merged**